cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,377 @@
+ */
+template <typename T>
+void compute_recurrence(rmm::device_uvector<thrust::pair<T, T>>& input,
+                        rmm::cuda_stream_view stream)
+{
+  recurrence_functor<T> op;
+  thrust::inclusive_scan(rmm::exec_policy(stream), input.begin(), input.end(), input.begin(), op);
+}
+
+/**
+* @brief Return an array whose values y_i are the number of null entries
+* in between the last valid entry of the input and the current index.
+* Example: {1, NULL, 3, 4, NULL, NULL, 7}
+        -> {0, 0     1, 0, 0,    1,    2}
+*/
+rmm::device_uvector<cudf::size_type> null_roll_up(column_view const& input,
+                                                  rmm::cuda_stream_view stream)
+{
+  rmm::device_uvector<cudf::size_type> output(
+    input.size(), stream, rmm::mr::get_current_device_resource());
+
+  auto device_view = *column_device_view::create(input);
+  auto valid_it    = cudf::detail::make_validity_iterator(device_view);
+
+  // TODO - not sure why two iterators produce a different result
+  // Invert the null iterator
+  thrust::transform(rmm::exec_policy(stream),
+                    valid_it,
+                    valid_it + input.size(),
+                    output.begin(),
+                    [=] __device__(bool valid) -> bool { return 1 - valid; });
+
+  // 0, 1, 0, 1, 1, 0 -> 0, 0, 1, 0, 0, 2
+  thrust::inclusive_scan_by_key(
+    rmm::exec_policy(stream), output.begin(), output.end() - 1, output.begin(), output.begin() + 1);
+
+  return output;
+}
+
+/**
+* @brief modify the source pairs that eventually yield the numerator
+* and denoninator to account for nan values. Pairs at nan indicies
+* become the identity operator (1, 0). The first pair after a nan
+* value or sequence of nan values has its first element multiplied by
+* N factors of beta, where N is the number of preceeding NaNs.
+*
+* This function requires that the null rollup be passed in as a
+* precomputed device vector. This is because, as in the case with
+* compute_ewma_noadjust, it is possible that the null rollup is needed
+* later in the calculation for something else, and we want to avoid
+* computing it twice.
+
+*/
+template <typename T>
+void pair_beta_adjust(column_view const& input,
+                      rmm::device_uvector<thrust::pair<T, T>>& pairs,
+                      rmm::device_uvector<cudf::size_type>& nullcnt,
+                      rmm::cuda_stream_view stream)
+{
+  using pair_type        = thrust::pair<T, T>;
+  auto device_view       = *column_device_view::create(input);
+  auto valid_it          = cudf::detail::make_validity_iterator(device_view);
+  auto valid_and_nullcnt = thrust::make_zip_iterator(thrust::make_tuple(valid_it, nullcnt.begin()));
+
+  thrust::transform(
+    rmm::exec_policy(stream),
+    valid_and_nullcnt,
+    valid_and_nullcnt + input.size(),
+    pairs.begin(),
+    pairs.begin(),
+    [=] __device__(thrust::tuple<bool, int> valid_and_nullcnt, pair_type pair) -> pair_type {
```

Comment:
I don't think there's anything that you need to capture here, but ignore this if I'm missing something (although it's probably better to specify exactly which captures you need if I am).
```suggestion
    [] __device__(thrust::tuple<bool, int> valid_and_nullcnt, pair_type pair) -> pair_type {
```